pkg/loki/modules.go

```diff
@@ -1378,6 +1372,66 @@ func (t *Loki) initQueryFrontend() (_ services.Service, err error) {
 	}), nil
 }
 
+func (t *Loki) initV2QueryEngine() (services.Service, error) {
+	if !t.Cfg.Querier.EngineV2.Enable {
+		return nil, nil
+	}
+
+	store, err := t.createDataObjBucket("query-engine")
+	if err != nil {
+		return nil, err
+	}
+
+	t.queryEngineV2 = &engine_v2.Engine{
+		Logger:     log.With(util_log.Logger, "module", "query-engine"),
+		Registerer: prometheus.DefaultRegisterer,
+
+		Config:          t.Cfg.Querier.EngineV2.Executor,
+		MetastoreConfig: t.Cfg.DataObj.Metastore,
+
+		Scheduler: t.queryEngineV2Scheduler,
+		Bucket:    store,
+		Limits:    t.Overrides,
+	}
+
+	return nil, nil
+}
+
+func (t *Loki) initV2QueryEngineScheduler() (services.Service, error) {
+	if !t.Cfg.Querier.EngineV2.Enable {
+		return nil, nil
+	}
+
+	sched := &engine_v2.Scheduler{
+		Logger: log.With(util_log.Logger, "module", "query-engine-scheduler"),
+	}
+
+	t.queryEngineV2Scheduler = sched
+	return sched.Service()
+}
+
+func (t *Loki) initV2QueryEngineWorker() (services.Service, error) {
+	if !t.Cfg.Querier.EngineV2.Enable {
+		return nil, nil
+	}
+
+	store, err := t.createDataObjBucket("query-engine-worker")
+	if err != nil {
+		return nil, err
+	}
+
+	worker := &engine_v2.Worker{
+		Logger: log.With(util_log.Logger, "module", "query-engine-worker"),
+		Bucket: store,
+
+		Config:   t.Cfg.Querier.EngineV2.Worker,
+		Executor: t.Cfg.Querier.EngineV2.Executor,
+
+		LocalScheduler: t.queryEngineV2Scheduler,
+	}
+	return worker.Service()
```

Comment:
As mentioned above, I think we should use existing conventions to construct services using constructor functions.